File: how_to_solve_coding_problems/sum_array_google.cc
#include <iostream>
#include <vector>
#include <unordered_set>

bool HasPairWithSum(const std::vector <int> data, int sum){
    // hash set in c++ is an unordered_set
    
    // unordered_set is an associative container that contains a 
    // set of unique objects of type Key
    // Search, insertion, and removal have average constant-time complexity.
    
    // constant-time complexity If an algorithm's time complexity is constant, 
    // it means that it will always run in the same amount of time, no matter the input size
    // O(1)
    
	std::unordered_set <int> comp; // complements
	for (int value: data){
	    // first check and then insert
	    // Check if I have seen it
		if (comp.find(value) != comp.end())
		    // find() return Iterator to an element with key equivalent to key. 
		    // If no such element is found, past-the-end (see end()) iterator is returned. 
		    // If the element is not found .end() is returned
		    // if we store find() value, with & we got addres and with * the value of the address
		    
            //.end() Returns an iterator to the 
            // element following the last element of the unordered_set
			return true;
		comp.insert(sum - value);
	}
	return false;
}

/*  
Collection of numbers, take the maching par that is 
equal to a sum that is given by the user.

[1, 2, 3, 4] sum = 8 -> no
[1, 2, 4, 4] sum = 8 -> yes
*/

int main(){
    std::vector <int> vect { 5, 2, 3, 9};
    std::vector <int> vect2 { 1, 2, 4, 4};
    int sum = 8;

    std::cout << HasPairWithSum(vect, sum) << std::endl;
    std::cout << HasPairWithSum(vect2, sum) << std::endl;
    
    return 0;
}

File: how_to_solve_coding_problems/sum_array_google.py
# Exercise
# Collection of numbers, take the maching par that is 
# equal to a sum that is given by the user.
# [1, 2, 3, 4] sum = 8 -> no
# [1, 2, 4, 4] sum = 8 -> yes

# Key points: one array not neccesarily sorted
# input: one array of any length and a integer
# output: boolean

# brute approach: nested loop
# complexity of O(n^2)

# smarter approach:
# Create a set with the complements
# check later if there is already the complemen in the set
# set.add worst case: O(n)
# search in a set worst case: O(n)

def find_pair_sum(arr, sum):
    complement = set()
    for number in arr:
        if number in complement:
            return True
        complement.add(sum - number)
    return False

def main():
    arr1 = [1, 3, 4, 5, 8, 12]
    # it works for no params and 0
    print(find_pair_sum(arr1, 8))


if __name__ == '__main__':
    main()